src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,450 @@
+	payload["key"] = str_key;
+
+	static const char* event_string[] = {"completed", "started", "stopped", "paused"};
+	if(req.event != event_t::none)
+		payload["event"] = event_string[static_cast<int>(req.event) - 1];
+
+	payload["peer_id"] = from_latin1({req.pid.data(), req.pid.size()});
+
+	payload["offers"] = json::array();
+	for(auto const& offer : req.offers)
+	{
+		json payload_offer;
+		payload_offer["offer_id"] = from_latin1({offer.id.data(), difference_type(offer.id.size())});
+		payload_offer["offer"]["type"] = "offer";
+		payload_offer["offer"]["sdp"] = offer.sdp;
+		payload["offers"].push_back(payload_offer);
+	}
+
+	std::string const data = payload.dump();
+
+#ifndef TORRENT_DISABLE_LOGGIN
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+#endif
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+    m_websocket->async_write(boost::asio::const_buffer(data.data(), data.size())
+     		, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_send(tracker_answer const& ans)
+{
+    json payload;
+    payload["action"] = "announce";
+    payload["info_hash"] = from_latin1({ans.info_hash.data(), difference_type(ans.info_hash.size())});
+    payload["offer_id"] = from_latin1({ans.answer.offer_id.data(), difference_type(ans.answer.offer_id.size())});
+    payload["to_peer_id"] = from_latin1({ans.answer.pid.data(), difference_type(ans.answer.pid.size())});
+    payload["peer_id"] =  from_latin1({ans.pid.data(), difference_type(ans.pid.size())});
+    payload["answer"]["type"] = "answer";
+    payload["answer"]["sdp"] = ans.answer.sdp;
+
+	std::string const data = payload.dump();
+
+#ifndef TORRENT_DISABLE_LOGGIN
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+#endif
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_write(boost::asio::const_buffer(data.data(), data.size())
+			, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_read()
+{
+	if(!m_websocket->is_open()) return;
+
+	// Can be replaced by m_read_buffer.clear() with boost 1.70+
+	if(m_read_buffer.size() > 0) m_read_buffer.consume(m_read_buffer.size());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_read(m_read_buffer
+            , std::bind(&websocket_tracker_connection::on_read, self, _1, _2));
+}
+
+void websocket_tracker_connection::on_connect(error_code const& ec)
+{
+	if(ec)
+	{
+#ifndef TORRENT_DISABLE_LOGGIN
+		std::shared_ptr<request_callback> cb = requester();
+		if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_FAILED [ url: %s, error: %d ]"
+				, tracker_req().url.c_str(), int(ec.value()));
+#endif
+		while(!m_pending.empty()) m_pending.pop();
```

Comment:
could you just do this instead? `m_pending = {};`

Reply:
Yes indeed, force of habit from old-fashioned C++ I guess...